### ISSUE TITLE
The mechanism of monitoring mountinfo should be optimized

### ISSUE THREAD

**Issue by syyhao1994** [systemd/systemd#34690]
### Is your feature request related to a problem? Please describe

https://github.com/systemd/systemd/issues/16728
https://github.com/systemd/systemd/issues/16728
Based on the above two problems, I think the current mountinfo monitoring mechanism is not perfect.
As long as enough mount points are mounted and the mountinfo is changing all the time, the performance of systemd becomes poor, and systemctl command are not even responded.

### Describe the solution you'd like

1.Can we  put the monitoring behavior of the mouninfo file  put into the subprocess or thread to avoid blocking the main systemd process?
2.Or we provide a config to limit the number of .mount can be processed in /proc/self/mountinfo.

**Comment by poettering** [comment]
> #16728
> #16728

You linked the same issue twice.

> Based on the above two problems, I think the current mountinfo monitoring mechanism is not perfect.

Yeah, not our fault. There has been a lon discussion about improving this in the kernel, and propagating mount notifications to userspace via fanotify. I think everyone there is on the same page at this time, but it hasn't been completed yet.

But again: fix our damn kubernetes first!

**Comment by poettering** [comment]
> 2.Or we provide a config to limit the number of .mount can be processed in /proc/self/mountinfo.

We put a ratelimit on the event source, if you look closely.

**Comment by syyhao1994** [comment]
> You linked the same issue twice.
> 

Sorry, i mean this PR as well.
https://github.com/systemd/systemd/pull/34614

> Yeah, not our fault. There has been a lon discussion about improving this in the kernel, and propagating mount notifications to userspace via fanotify. I think everyone there is on the same page at this time, but it hasn't been completed yet.

Yes, i found we have discussed this problem in 2020. But i think there is a long way to go before the kernel provides the method to do this. So i think whether we could do the optimize in systemd. As the problem will occure when there are to many mount points regardless k8s or kata.

> We put a ratelimit on the event source, if you look closely.
Yes, i now that. But if there are too many mount points(maybe 5000+), it takes a few seconds to process an mountinfo notify event.
So i come up with two ideas mentioned above about this problem.

>1.Can we put the monitoring behavior of the mouninfo file put into the subprocess or thread to avoid blocking the main systemd process?
I am not sure if this idea can be realized. 

>2.Or we provide a config to limit the number of .mount can be processed in /proc/self/mountinfo.
I mean we only process maybe 2000 or 3000 mount units in /proc/self/mountinfo at most to control the load of systemd.

**Comment by yuwata** [comment]
#34614 is merged. Let's see if it helps. Closing.

**Comment by syyhao1994** [comment]
> #34614 is merged. Let's see if it helps. Closing.

I notice this, but this could only optimize a little rather than solve this.
What i want to discuss is how we can do to optimize the mothod of process mounintfo to make sure systemd will not process mounintfo all the time as the priority of mount events is realy high no matter how many mountinfos come up.

**Comment by yuwata** [comment]
OK, but I think further optimization should be done in k8s side.

**Comment by bluca** [comment]
> > Based on the above two problems, I think the current mountinfo monitoring mechanism is not perfect.
> 
> Yeah, not our fault. There has been a lon discussion about improving this in the kernel, and propagating mount notifications to userspace via fanotify. I think everyone there is on the same page at this time, but it hasn't been completed yet.

Wasn't this implemented very very recently? @brauner ?

**Comment by syyhao1994** [comment]
> OK, but I think further optimization should be done in k8s side.

Yes, and kata has solved this in https://github.com/kata-containers/kata-containers/pull/8760/commits/9317e23df16dd945984ab8af8f1a875dc3303983.
But i think the way we use for processing the mountinfo could be optimized.For example, if someone maliciously creates a large number of mount points, then DOS provisioning may occur

**Comment by syyhao1994** [comment]
https://github.com/torvalds/linux/commit/46eae99ef73302f9fb3dddcd67c374b3dffe8fd6#diff-26b064824bfafef546f17235dc1f34dbea87ce6c27f9320b5d4427bfa4d2b0e9R4933
Maybe some optimize could be done with statemount？

**Comment by JetXujing** [comment]
When there are many mount points (for example, 6000 mount points), the elapsed time is collected. The elapsed time is mainly caused by unit_remove_dependencies. This function is introduced by https://github.com/systemd/systemd/commit/38b9b72ec10fa8875b1b7c93bc7ca6b8a2d5486e, I am wondering if it is possible to remove unit_remove_dependencies to reduce the time. In my preliminary opinion, deleting this function will only add more dependencies, which is more acceptable than performance degradation.

**Comment by JetXujing** [comment]
In extreme cases, this can make various systemd-related commands unavailable (causing command execution to time out).

**Comment by bluca** [comment]
> [torvalds/linux@46eae99#diff-26b064824bfafef546f17235dc1f34dbea87ce6c27f9320b5d4427bfa4d2b0e9R4933](https://github.com/torvalds/linux/commit/46eae99ef73302f9fb3dddcd67c374b3dffe8fd6#diff-26b064824bfafef546f17235dc1f34dbea87ce6c27f9320b5d4427bfa4d2b0e9R4933) Maybe some optimize could be done with statemount？

That's the first part, this is what I was referring to: https://lwn.net/Articles/980330/
I'm not sure that work has landed yet. When it does, yes we should use it and it should make the performance issues go away. We can't just remove needed functionality to improve performance edge cases, sorry. Once that kernel functionality is available, happy to review patches that make use of it instead of the manual proc parsing.

**Comment by bluca** [comment]
> [torvalds/linux@46eae99#diff-26b064824bfafef546f17235dc1f34dbea87ce6c27f9320b5d4427bfa4d2b0e9R4933](https://github.com/torvalds/linux/commit/46eae99ef73302f9fb3dddcd67c374b3dffe8fd6#diff-26b064824bfafef546f17235dc1f34dbea87ce6c27f9320b5d4427bfa4d2b0e9R4933) Maybe some optimize could be done with statemount？

And as you mentioned, maybe doing busy polling with statmount instead of parsing proc could already provide some benefits, if someone were to work on that we'd be happy to review it, if it works and provides benefits

**Comment by Forrest-ly** [comment]
@bluca @poettering Linux kernel & libmount seem to already support “mount notification”. Can systemd also support it?
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?spm=149fea8c.172eb9e6.0.0.4ac56fc0ilhDYX&h=v6.17-rc7&id=2cc0b7fd4bb0cd7f98fe75758e4c619f74873bd9

https://github.com/util-linux/util-linux/commit/31a9fc781e92501a98f521c74845c333547989ac?spm=149fea8c.172eb9e6.0.0.4ac56fc0ilhDYX
